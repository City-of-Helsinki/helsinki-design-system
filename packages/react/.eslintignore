**/*.d.ts
webpack.config.js
.templates
lib/
**/*.stories.tsx
