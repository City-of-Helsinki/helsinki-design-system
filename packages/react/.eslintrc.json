{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:react/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:jest/recommended",
    "airbnb",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended" // this should always be the last item on this array
  ],
  "plugins": ["react", "@typescript-eslint", "prettier", "jest", "import"],
  "parserOptions": {
    "ecmaVersion": 2018,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "ignorePatterns": ["*.module.css"],
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error"],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*.test.ts", "**/*.test.tsx", "**/test*.ts*", "**/*.stories.tsx"]
      }
    ],
    "import/order": [
      "error",
      {
        "groups": ["builtin", "external", "internal", ["parent", "sibling", "index"]],
        "newlines-between": "always"
      }
    ],
    "react/jsx-filename-extension": "off",
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-props-no-spreading": "off",
    "react/jsx-curly-newline": "off",
    "@typescript-eslint/explicit-function-return-type": "off",
    "react/prop-types": "off",
    "import/extensions": "off",
    "import/prefer-default-export": "off",
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "components": ["Label"],
        "required": {
          "some": ["nesting", "id"]
        },
        "allowChildren": false
      }
    ],
    "@typescript-eslint/explicit-member-accessibility": "off",
    "react/state-in-constructor": [2, "never"],
    "no-unused-expressions": [2, {
      "allowTernary": true
    }],
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": ["error"]
  },
  "settings": {
    "react": {
      "version": "detect"
    },
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx"],
        "paths": ["src"]
      }
    }
  },
  "env": {
    "browser": true,
    "jest": true
  },
  "globals": {
    "JSX": true
  }
}
