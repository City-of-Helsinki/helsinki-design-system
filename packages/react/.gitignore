node_modules/
storybook-static/
lib/
_bundles/
*.css.d.ts
.loki/current
.loki/difference

*storybook.log
