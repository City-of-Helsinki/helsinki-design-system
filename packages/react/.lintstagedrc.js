module.exports = {
  'src/**/*.{ts,tsx}': 'eslint',
}
