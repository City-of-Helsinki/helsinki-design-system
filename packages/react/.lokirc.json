{
  "chromeRetries": 2,
  "configurations": {
    "chrome.laptop": {
      "target": "chrome.docker",
      "width": 1366,
      "height": 768,
      "mobile": false
    },
    "chrome.iphone7": {
      "target": "chrome.docker",
      "preset": "iPhone 7"
    }
  },
  "skipStories": "Icons/.*|Button/*|Tag/*|Hero/*|Koros/*|Checkbox/*",
  "pixelmatch": {
    "threshold": 0.2,
    "alpha": 1,
    "diffColor": [255, 0, 255]
  }
}
