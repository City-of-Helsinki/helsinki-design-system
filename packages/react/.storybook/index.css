@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/565d73a693abe0776c801607ac28f0bf.woff') format('woff');
  font-display: swap;
  font-weight: 400;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/5bb29e3b7b1d3ef30121229bbe67c3e1.woff') format('woff');
  font-display: swap;
  font-weight: 400;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/7c46f288e8133b87e6b12b45dac71865.woff') format('woff');
  font-display: swap;
  font-weight: 500;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/e62dc97e83a385e4d8cdc939cf1e4213.woff') format('woff');
  font-display: swap;
  font-weight: 500;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/533af26cf28d7660f24c2884d3c27eac.woff') format('woff');
  font-display: swap;
  font-weight: 700;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/20d494430c87e15e194932b729d48270.woff') format('woff');
  font-display: swap;
  font-weight: 700;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/a50a1bd245ce63abcc0d1da80ff790d2.woff') format('woff');
  font-display: swap;
  font-weight: 900;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/62a1781d8b396fbb025b0552cf6304d2.woff') format('woff');
  font-display: swap;
  font-weight: 900;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

body {
  font-family: var(--font-default);
  font-size: var(--fontsize-body-m);
  color: var(--color-black-90);
  line-height: var(--lineheight-l);
}

.password-input--external-show-password-button,
.date-input--external-clear-value-button {
  display: grid;
  grid-template-columns: 400px 200px;
  gap: 20px;
  align-items: center;
}

@media only screen and (max-width: 768px) {
  .password-input--external-show-password-button,
  .date-input--external-clear-value-button {
    grid-template-columns: 300px;
  }
}

@media only screen and (max-width: 575.98px) {
  .table-custom-action {
    width: 100%;
  }
}

/* STEPPER */

.stepper-small {
  max-width: 300px;
}

@media only screen and (max-width: 575.98px) {
  .stepper-small {
    max-width: 100%;
  }
}

.stepper-margin {
  margin-left: -35px;
}

.stepper-heading {
  margin-left: 10px;
}

.stepper-form-validation {
  margin-left: -18px;
}

.stepper-card > div:first-child {
  margin-bottom: var(--spacing-m) !important;
}

#focused-element-after-cookie-consent-closed:focus {
  outline: 2px black solid;
}
