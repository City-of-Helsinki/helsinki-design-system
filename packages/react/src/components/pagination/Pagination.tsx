import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { VisuallyHidden } from '@react-aria/visually-hidden';

import '../../styles/base.module.css';
import styles from './Pagination.module.scss';
import classNames from '../../utils/classNames';
import { Button, ButtonPresetTheme, ButtonVariant } from '../button';
import { IconAngleLeft, IconAngleRight } from '../../icons';
import { useTheme } from '../../hooks/useTheme';
import { AllElementPropsWithoutRef, MergeAndOverrideProps } from '../../utils/elementTypings';

type Language = 'en' | 'fi' | 'sv';

const mapLangToNext = (language: string): string => {
  const next = {
    fi: 'Seuraava',
    sv: 'Nästa',
    en: 'Next',
  };

  return next[language];
};

const mapLangToPrevious = (language: string): string => {
  const previous = {
    fi: 'Edellinen',
    sv: 'Föregående',
    en: 'Previous',
  };

  return previous[language];
};

const mapLangToPageAriaLabel = (pageNumber: number, language: Language): string => {
  const pageAriaLabel = {
    fi: `Sivu ${pageNumber}`,
    sv: `Sida ${pageNumber}`,
    en: `Page ${pageNumber}`,
  };

  return pageAriaLabel[language];
};

const mapLangToPageTitle = (pageNumber: number, language: Language, currentPage: boolean): string => {
  const currentPageTitle = {
    fi: 'Nykyinen sivu',
    sv: 'Nuvarande sida',
    en: 'Current page',
  };

  if (currentPage) {
    return currentPageTitle[language];
  }

  const pageTitle = {
    fi: `Sivu ${pageNumber}`,
    sv: `Sida ${pageNumber}`,
    en: `Page ${pageNumber}`,
  };

  return pageTitle[language];
};

const mapLangToOpenedPage = (pageNumber: number, language: Language): string => {
  const openedPage = {
    fi: `Avattu sivu ${pageNumber}`,
    sv: `Öppnad sida ${pageNumber}`,
    en: `Opened page ${pageNumber}`,
  };

  return openedPage[language];
};

const range = (start: number, end: number): number[] => {
  const length = end - start + 1;
  return Array.from({ length }, (_, i) => start + i);
};

enum Ellipsis {
  start = 'start-ellipsis',
  end = 'end-ellipsis',
}

const createPaginationItemList = ({
  pageCount,
  pageIndex,
  siblingCount,
}: Partial<PaginationProps>): (Ellipsis | number)[] => {
  const endPages = range(Math.max(pageCount, 2), pageCount);
  const startPages = range(1, Math.min(1, pageCount));

  const siblingsStart = Math.max(
    Math.min(
      // Natural start
      pageIndex + 1 - siblingCount,
      // Lower boundary when page is high
      pageCount - 1 - siblingCount * 2 - 1,
    ),
    // Greater than startPages, because we will insert startPages separately
    3,
  );

  const siblingsEnd = Math.min(
    Math.max(
      // Natural end
      pageIndex + 1 + siblingCount,
      // Upper boundary when page is low
      1 + siblingCount * 2 + 2,
    ),
    // Less than endPages
    endPages.length > 0 ? endPages[0] - 2 : pageCount - 1,
  );

  return [
    ...startPages,
    // start ellipsis
    // eslint-disable-next-line no-nested-ternary
    ...(siblingsStart > 3 ? [Ellipsis.start] : pageCount - 1 > 2 ? [2] : []),

    // Sibling pages
    ...range(siblingsStart, siblingsEnd),

    // End ellipsis
    // eslint-disable-next-line no-nested-ternary
    ...(siblingsEnd < pageCount - 2 ? [Ellipsis.end] : pageCount - 1 > 1 ? [pageCount - 1] : []),

    ...endPages,
  ];
};

export interface PaginationCustomTheme {
  /**
   * Colour of the active page background. Use black, white or any of the brand colours.
   */
  '--active-page-background-color'?: string;
}

export type PaginationProps = MergeAndOverrideProps<
  AllElementPropsWithoutRef<'nav'>,
  {
    /**
     * If true, hide the next-page button
     * @default false
     */
    hideNextButton?: boolean;
    /**
     * If true, hide the previous-page button
     * @default false
     */
    hidePrevButton?: boolean;
    /**
     * The language of the pagination component.
     * @default fi
     */
    language?: Language;
    /**
     * Callback fired when the page is changed
     */
    onChange?: (
      event: React.MouseEvent<HTMLAnchorElement> | React.MouseEvent<HTMLButtonElement>,
      index: number,
    ) => void;
    /**
     * The total number of pages
     */
    pageCount: number;
    /**
     * A function for generating the href of pages
     */
    pageHref: (index: number) => string;
    /**
     * The active page index
     */
    pageIndex: number;
    /**
     * Aria-label for the pagination nav element
     */
    paginationAriaLabel: string;
    /**
     * Number of always visible pages before and after the current page
     * @default 1
     */
    siblingCount?: number;
    /**
     * Theme prop for customisation of the Pagination component
     */
    theme?: PaginationCustomTheme;
  }
>;

export const Pagination = ({
  'data-testid': dataTestId,
  hideNextButton = false,
  hidePrevButton = false,
  language = 'fi',
  onChange,
  pageCount,
  pageHref,
  pageIndex,
  paginationAriaLabel,
  siblingCount = 1,
  theme,
  className,
  ...rest
}: PaginationProps) => {
  const initialPageIndex = useRef(pageIndex);
  const activeItemRef = useRef<HTMLSpanElement>();
  const userSelectedIndex = useRef(-1);
  const [hasUserChangedPage, setHasUserChangedPage] = useState<boolean>(false);

  const onChangeWithInternalHandler = useCallback(
    (event, index, wasButtonClick = false) => {
      // do not shift focus away from buttons
      userSelectedIndex.current = wasButtonClick ? -1 : index;
      if (onChange) {
        onChange(event, index);
      }
    },
    [onChange],
  );

  useEffect(() => {
    if (hasUserChangedPage === false) {
      if (pageIndex !== initialPageIndex.current) {
        setHasUserChangedPage(true);
      }
    }
  }, [pageIndex, hasUserChangedPage]);

  useEffect(() => {
    if (userSelectedIndex.current > -1 && activeItemRef.current) {
      // Active element changes from <a> to <span>, so focus is lost after re-render.
      // Move it back manually
      activeItemRef.current.focus();
    }
  }, [userSelectedIndex.current]);

  const itemList = useMemo(
    () => createPaginationItemList({ pageCount, pageIndex, siblingCount }),
    [pageCount, pageIndex, siblingCount],
  );

  const customThemeClass = useTheme<PaginationCustomTheme>(styles.pagination, theme);

  if (pageCount <= 0) {
    return null;
  }

  return (
    <div className={styles.container}>
      <nav
        {...rest}
        className={classNames(
          styles.pagination,
          customThemeClass,
          hideNextButton ? styles.hideNextButton : '',
          className,
        )}
        role="navigation"
        aria-label={paginationAriaLabel}
        data-next={mapLangToNext(language)}
        data-testid={dataTestId}
      >
        <VisuallyHidden>
          <span aria-atomic aria-live="polite">
            {hasUserChangedPage ? mapLangToOpenedPage(pageIndex + 1, language) : ''}
          </span>
        </VisuallyHidden>
        {!hidePrevButton && (
          <Button
            className={styles.buttonPrevious}
            data-testid={dataTestId ? `${dataTestId}-previous-button` : undefined}
            disabled={pageIndex === 0 || pageCount === 1}
            aria-disabled={pageIndex === 0 || pageCount === 1 || undefined}
            onClick={(event) => onChangeWithInternalHandler(event, pageIndex - 1, true)}
            variant={ButtonVariant.Supplementary}
            theme={ButtonPresetTheme.Black}
            iconStart={<IconAngleLeft />}
          >
            {mapLangToPrevious(language)}
          </Button>
        )}
        <ul className={styles.pages}>
          {itemList.map((pageItem) => {
            if (pageItem === Ellipsis.start || pageItem === Ellipsis.end) {
              return (
                <li key={pageItem}>
                  <span className={styles.itemEllipsis}>...</span>
                </li>
              );
            }

            const isCurrent = pageIndex + 1 === pageItem;

            return (
              <li key={pageItem}>
                {isCurrent ? (
                  <span
                    className={classNames(styles.itemLink, styles.itemLinkActive)}
                    data-testid={dataTestId ? `${dataTestId}-page-${pageItem}` : undefined}
                    aria-label={`${mapLangToPageAriaLabel(pageItem, language)}. ${mapLangToPageTitle(pageItem, language, true)}.`}
                    aria-current="page"
                    tabIndex={-1}
                    ref={activeItemRef}
                  >
                    {pageItem}
                  </span>
                ) : (
                  <a
                    className={classNames(styles.itemLink)}
                    data-testid={dataTestId ? `${dataTestId}-page-${pageItem}` : undefined}
                    href={pageHref(pageItem)}
                    onClick={(event) => onChangeWithInternalHandler(event, pageItem - 1)}
                    title={mapLangToPageTitle(pageItem, language, false)}
                    aria-label={mapLangToPageAriaLabel(pageItem, language)}
                  >
                    {pageItem}
                  </a>
                )}
              </li>
            );
          })}
        </ul>
        {!hideNextButton && (
          <Button
            className={styles.buttonNext}
            data-testid={dataTestId ? `${dataTestId}-next-button` : undefined}
            disabled={pageIndex === pageCount - 1 || pageCount === 1}
            aria-disabled={pageIndex === pageCount - 1 || pageCount === 1 || undefined}
            onClick={(event) => onChangeWithInternalHandler(event, pageIndex + 1, true)}
            variant={ButtonVariant.Supplementary}
            theme={ButtonPresetTheme.Black}
            iconEnd={<IconAngleRight className={styles.angleRightIcon} />}
          >
            {mapLangToNext(language)}
          </Button>
        )}
      </nav>
    </div>
  );
};
