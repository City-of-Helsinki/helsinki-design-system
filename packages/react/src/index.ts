export * from './icons';
export * from './components';
export * from './ssr';
