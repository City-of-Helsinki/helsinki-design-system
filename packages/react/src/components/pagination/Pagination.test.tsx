import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { Pagination } from './Pagination';

describe('<Pagination /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(
      <Pagination
        onChange={() => null}
        paginationAriaLabel="Pagination"
        pageCount={68}
        pageIndex={7}
        pageHref={() => '#'}
        language="en"
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(
      <Pagination
        onChange={() => null}
        paginationAriaLabel="Pagination"
        pageCount={68}
        pageIndex={7}
        pageHref={() => '#'}
        language="en"
      />,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
