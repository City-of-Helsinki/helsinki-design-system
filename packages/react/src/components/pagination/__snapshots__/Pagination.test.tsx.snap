// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <nav
      aria-label="Pagination"
      class="pagination"
      data-next="Next"
      role="navigation"
    >
      <button
        class="button supplementary theme-black size-default buttonPrevious"
        type="button"
      >
        <div
          aria-hidden="true"
          class="icon"
        >
          <svg
            class="icon s"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <polygon
                points="24 0 24 24 0 24 0 0"
              />
              <polygon
                fill="currentColor"
                points="10.5 12 15.5 17 14 18.5 7.5 12 14 5.5 15.5 7"
              />
            </g>
          </svg>
        </div>
        <span
          class="label"
        >
          Previous
        </span>
      </button>
      <ul
        class="pages"
      >
        <li>
          <a
            aria-current="false"
            aria-label="Page 1"
            class="itemLink"
            href="#"
            title="Page 1"
          >
            1
          </a>
        </li>
        <li>
          <span
            class="itemEllipsis"
          >
            ...
          </span>
        </li>
        <li>
          <a
            aria-current="false"
            aria-label="Page 7"
            class="itemLink"
            href="#"
            title="Page 7"
          >
            7
          </a>
        </li>
        <li>
          <a
            aria-current="page"
            aria-label="Page 8"
            class="itemLink itemLinkActive"
            href="#"
            title="Current page"
          >
            8
          </a>
        </li>
        <li>
          <a
            aria-current="false"
            aria-label="Page 9"
            class="itemLink"
            href="#"
            title="Page 9"
          >
            9
          </a>
        </li>
        <li>
          <span
            class="itemEllipsis"
          >
            ...
          </span>
        </li>
        <li>
          <a
            aria-current="false"
            aria-label="Page 68"
            class="itemLink"
            href="#"
            title="Page 68"
          >
            68
          </a>
        </li>
      </ul>
      <button
        class="button supplementary theme-black size-default buttonNext"
        type="button"
      >
        <span
          class="label"
        >
          Next
        </span>
        <div
          aria-hidden="true"
          class="icon"
        >
          <svg
            class="icon s angleRightIcon"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <polygon
                points="0 24 0 0 24 0 24 24"
              />
              <polygon
                fill="currentColor"
                points="13.5 12 8.5 7 10 5.5 16.5 12 10 18.5 8.5 17"
              />
            </g>
          </svg>
        </div>
      </button>
    </nav>
  </div>
</DocumentFragment>
`;
